src/main/scala/sparkz/core/network/peer/PeerManager.scala

```diff
@@ -64,23 +64,24 @@ class PeerManager(settings: SparkzSettings, sparkzContext: SparkzContext) extend
             log.info(s"New discovered peer: $peerInfo")
             peerInfo
         }
-      peerDatabase.addOrUpdateKnownPeers(filteredPeers)
+      peerDatabase.addOrUpdateKnownPeers(filteredPeers, Some(source))
 
     case RemovePeer(address) =>
-      if (!knownPeersSet(address)) {
-        peerDatabase.remove(address)
-        log.info(s"$address removed from peers database")
-      }
+      peerDatabase.remove(address)
+      log.info(s"$address removed from peers database")
+
+    case get: RandomPeerForConnectionExcluding =>
+      sender() ! get.choose(knownPeers, peerDatabase.randomPeersSubset, peerDatabase.blacklistedPeers, sparkzContext)
 
     case get: GetPeers[_] =>
-      sender() ! get.choose(peerDatabase.knownPeers, peerDatabase.blacklistedPeers, sparkzContext)
+      sender() ! get.choose(knownPeers, peerDatabase.allPeers, peerDatabase.blacklistedPeers, sparkzContext)
   }
 
   private def apiInterface: Receive = {
 
     case GetAllPeers =>
-      log.trace(s"Get all peers: ${peerDatabase.knownPeers}")
-      sender() ! peerDatabase.knownPeers
+      log.trace(s"Get all peers: ${peerDatabase.allPeers}")
+      sender() ! peerDatabase.allPeers
```

Comment:
GetAllPeers is now missing knownPeers.
Also, looking at `private def apiInterface: Receive = {` it seems that it's never reached as all its cases are covered by
`case get: GetPeers[_] => ` at line 75/76 which is evaluated earlier

Reply:
Yes, you're right, that is something I've noticed when writing the networking layer doc for the old scorex tag. I think we can remove the apiInterface case since it can all be handled by the [general case](https://github.com/HorizenOfficial/Sparkz/blob/main/src/main/scala/sparkz/core/network/peer/PeerManager.scala#L74)

Reply:
what about this one?